modules/em/src/BayesEM3D.cpp

```diff
@@ -0,0 +1,532 @@
+  return bayesem3d_get_logsumexp(x + log(wx), y + log(wy));
+}
+
+std::vector<double> bayesem3d_get_value(Particle *p,
+                                        const algebra::Vector3D &pt,
+                                        double mass_ii,
+                                        const IMP::em::KernelParameters &kps) {
+  std::vector<double> results;
+  algebra::Vector3D cs = core::XYZ(p).get_coordinates();
+
+  double rsq = (cs - pt).get_squared_magnitude();
+  double norm = (kps.get_rnormfac() * mass_ii);
+
+  double tmp = -rsq * kps.get_inv_rsigsq();
+  results.push_back(tmp);
+  results.push_back(norm);
+
+  results.push_back(-2. * (cs[0] - pt[0]) * kps.get_inv_rsigsq());
+  results.push_back(-2. * (cs[1] - pt[1]) * kps.get_inv_rsigsq());
+  results.push_back(-2. * (cs[2] - pt[2]) * kps.get_inv_rsigsq());
+
+  return results;
+}
+
+std::vector<double> bayesem3d_get_value_no_deriv(
+    Particle *p, const algebra::Vector3D &pt, double mass_ii,
+    const IMP::em::KernelParameters &kps) {
+  std::vector<double> results;
+  algebra::Vector3D cs = core::XYZ(p).get_coordinates();
+
+  double rsq = (cs - pt).get_squared_magnitude();
+  double norm = (kps.get_rnormfac() * mass_ii);
+
+  double tmp = -rsq * kps.get_inv_rsigsq();
+  results.push_back(tmp);
+  results.push_back(norm);
+
+  return results;
+}
+
+FloatPair bayesem3d_get_em_density_squared(DensityMap *em,
+                                           long number_of_voxels) {
+  FloatPair results;
+
+  const double *em_data = em->get_data();
+  double em_square = 0;
+  double scale = 0;
+
+  for (long jj = 0; jj < number_of_voxels; ++jj) {
+    em_square += em_data[jj] * em_data[jj];
+    scale += em_data[jj];
+  }
+  results.first = scale;
+  results.second = em_square;
+  return results;
+}
+
+double bayesem3d_get_cross_correlation_coefficient(DensityMap *em1,
+                                                   DensityMap *em2) {
+  const DensityHeader *em1_header = em1->get_header();
+  const double *em1_data = em1->get_data();
+
+  const DensityHeader *em2_header = em2->get_header();
+  const double *em2_data = em2->get_data();
+
+  long nvox1 = em1_header->get_number_of_voxels();
+  long nvox2 = em2_header->get_number_of_voxels();
+
+  if (nvox1 != nvox2)
+    IMP_THROW("Can't compute correlation for different size maps.",
+              ValueException);
+
+  double em1_mean = 0.;
+  double em2_mean = 0.;
+
+  double em1_sqr = 0.;
+  double em2_sqr = 0.;
+
+  double cross_term = 0.;
+
+  for (long i = 0; i < nvox1; ++i) {
+    if (em1_data[i] > 0 && em2_data[i] > 0) {
+      em1_mean += em1_data[i];
+      em2_mean += em2_data[i];
+
+      em1_sqr += IMP::square(em1_data[i]);
+      em2_sqr += IMP::square(em2_data[i]);
+
+      cross_term += em1_data[i] * em2_data[i];
+    }
+  }
+
+  em1_mean /= nvox1;
+  em2_mean /= nvox2;
+
+  em1_sqr /= nvox1;
+  em2_sqr /= nvox2;
+
+  double sig1 = em1_sqr - IMP::square(em1_mean);
+  sig1 = std::max(0., sqrt(sig1));
```

Comment:
Doesn't `sqrt` always return a positive number anyway?

Reply:
Reverting to old version of code here for those two line:
Line: 252: `sig1 = ( sig1 <0. ? 0. : sqrt(sig1) );`
Line: 255: `sig2 = ( sig2 <0. ? 0. : sqrt(sig2) );`